### ISSUE TITLE
BEEFY: disaster-recovery: losing validator keys

### ISSUE THREAD

**Issue by acatangiu** [paritytech/substrate#14500]
With BEEFY there is the concern of enough validators losing their session/authority keys and thus BEEFY getting stuck.

The probability of this happening is **very low**, but impact is **critical** in case of trustless bridges based on BEEFY light-clients.

## Example scenarios with probability:
- AWS or GCP have a massive, even temporary, outage impacting 1/3 of active validators - if they lose ephemeral session keys during this event, consensus is broken - probability **very low**,
- assuming validators rotate session keys often (which they don't currently, but theoretically could/should), say BEEFY finality stops because of some bug (but GRANDPA continues ok) and it takes multiple sessions until we deploy a fix, it could happen that validators simply rotate their ephemeral keys out of existence (there is no mechanism preventing this rn) - probability **low**.

## Impact

- On Polkadot/Kusama (local) chain, this can be fixed through governance (or some other sudo) - **impact high**, but fixable
- In case of a trustless bridge like Polkadot <> Ethereum, where we want the bridge to be completely decentralized and trustless and be controlled/controllable only by Polkadot Governance, there is no fix: if the last active validator set cannot reach consensus, any governance intention/decision/action on Polkadot cannot be bridged over and acted upon on Ethereum; basically the BEEFY light-client on Ethereum side cannot be updated - **impact fatal**, bridge is dead, funds are lost (locked forever).

### On-chain encrypted backups for session keys (fix on Polkadot side)

Validators use a master key (can be their controller or stash key - which is likely very much not ephemeral, but on some hardware device, or KMS in case of controller keys, or similar) to encrypt the private key part of their ephemeral session keys pair and post the encrypted backup on-chain where it can be recovered.

Cons: "breaking" change, adds complexity to validators

### BEEFY light-client `ForceChangeAuthorities` support (fix on Ethereum side)

BEEFY light-client has fallback `ForceChangeAuthorities` mechanism to "force" new authorities.

One idea for implementing this is using GRANDPA keys or even controller keys of the authority set active in the last bridged header to sign a Smart Contract upgrade "custom" payload (either setting new BEEFY keys, or upgrading light-client, etc).
Validity/authenticity of the GRANDPA or controller keys signing this "force" change can be proven on ethereum side using Polkadot storage proofs verified against last bridged header `state_root`.

Cons:
- a lot of complexity on light-client side to support this "force" upgrade path,
- very expensive (doesn't really matter in a disaster scenario) - verifying Polkadot storage proofs in EVM, verifying `sr25519` signatures in EVM - would it even fit an eth block?

### Others?

cc @burdges, @vgeddes, @andresilva

**Comment by vgeddes** [comment]
Thanks for writing this up! Is definitely a tricky problem to solve. 

Some notes from my side:
* @seunlanlege developed a Solidity prover for substrate storage proofs here: https://github.com/polytope-labs/solidity-merkle-trees#merkle-patricia-trie. I vaguely recall the gas cost for verifying a proof was around 120,000, which isn't too bad.

**Comment by seunlanlege** [comment]
Isn't this already a risk with GRANDPA/BABE session keys which would first devastate the network's liveness

**Comment by acatangiu** [comment]
> Isn't this already a risk with GRANDPA/BABE session keys which would first devastate the network's liveness?

Yes, same risk there, but impact is not as bad as governance can `ForceSetChange` GRANDPA/BABE authorities.
In case of bridge-scenario, governance cannot act on the Eth side of the bridge while bridge is not functioning.

**Comment by seunlanlege** [comment]
> In case of bridge-scenario, governance cannot act on the Eth side of the bridge while bridge is not functioning.

Contracts generally will have admin keys or some kind of admin functionality, This is entirely for safety reasons which this scenario would fall under.

**Comment by acatangiu** [comment]
> Contracts generally will have admin keys or some kind of admin functionality, This is entirely for safety reasons which this scenario would fall under.

We are pursuing an architecture where this is not the case: fully trustless with no admin other than chain governance.

**Comment by burdges** [comment]
We do not various options here but.. 

It's likely simplest to do a backup encrypted to the controller, or encrypted to the controller plus its stash, or some other combination.  We'd need to support controllers and stashes in all combinations of SR25519, Ed25519, and secp256k1 yes?

**Comment by acatangiu** [comment]
> It's likely simplest to do a backup encrypted to the controller, or encrypted to the controller plus its stash, or some other combination.

On-chain encrypted backup, or did you have smth else in mind?

> We'd need to support controllers and stashes in all combinations of SR25519, Ed25519, and secp256k1 yes?

yes, I don't see why we'd do only a subset

**Comment by burdges** [comment]
An on-chain backup maybe easier, but we could store off-chain however folks like

Actually I've an amusing idea:  

A BEEFY node creates secp256k1 and Ristretto/Ed25519 secret keys, called `x` and `y`, with which it computes key exchanges with all other validator's controller keys, which produces 3f+1 = 1000ish random `[u8; 32]`s.  It treats these as scalars in the F_r field for BLS12-381, creates a Lagrange polynomial f that interpolates the first 3f+1 = 664ish of these as f(1), f(2), etc, and then creates a string M of the f = 333ish bitmasks that make the others fit the polynomial.  It then treats the evaluation z = f(0) as the BLS12-381 secret key.

Now finally the BEEFY node computes the public keys (X,Y1,Y2) for x and y, makes up an ephemeral key pair (u,U) of the key type as the controller public key C, computes the key exchange k = u * C, and forms a ciphertext Z, U, X, Y1, Y2, AEAD_k(M).

What happens if BEEFY keys are lost?  A controller key owner can download decrypt and validate their Z, X, Y1, Y2, M, so they can then ask 2/3 of the other controller key holders to compute key exchanges with whichever of X, Y1, Y2 fits them.   After these are returned then it reinterpolates f and recomputes z = f(0).

Is this better than simply encrypting to the controller?  It ensures a compromised controller cannot decrypt the BEEFY secret key, but one expect controllers would be air gaped, so one being compromised sounds unlikely.  And both controller and stash being compromised sound even more unlikely.  It avoids deployment complexity, but it makes the recovery pretty nasty, so meh.
